src/plone/app/theming/policy.py

```diff
@@ -0,0 +1,157 @@
+        except KeyError:
+            return None
+        return settings
+
+    def getCurrentTheme(self):
+        """The name of the current theme."""
+        settings = self.getSettings()
+        if not settings.rules:
+            return None
+        if settings.currentTheme:
+            return settings.currentTheme
+
+        # BBB: If currentTheme isn't set, look for a theme with a rules file
+        # matching that of the current theme
+        for theme in utils.getAvailableThemes():
+            if theme.rules == settings.rules:
+                return theme.__name__
+
+        return None
+
+    def isThemeEnabled(self, settings=None):
+        """Whether theming is enabled."""
+
+        # Resolve DevelopmentMode late (i.e. not on import time) since it may
+        # be set during import or test setup time
+        DevelopmentMode = Globals.DevelopmentMode
+
+        # Disable theming if the response sets a header
+        if self.request.response.getHeader('X-Theme-Disabled'):
+            return False
+
+        # Check for diazo.off request parameter
+        true_vals = ('1', 'y', 'yes', 't', 'true')
+        if (DevelopmentMode and self.request.get(
+                'diazo.off', '').lower() in true_vals):
+            return False
+
+        if not settings:
+            settings = self.getSettings()
+        if not settings.enabled or not settings.rules:
+            return False
+
+        server_url = self.request.get('SERVER_URL')
+        proto, host = server_url.split('://', 1)
+        host = host.lower()
+        serverPort = self.request.get('SERVER_PORT')
+
+        for hostname in settings.hostnameBlacklist or ():
+            if host == hostname or host == ':'.join((hostname, serverPort)):
+                return False
+
+        return True
+
+    def getCache(self, theme=None):
+        """Managing the cache is a policy decision."""
+        caches = self.getCacheStorage()
+        key = self.getCacheKey(theme)
+        cache = caches.get(key)
+        if cache is None:
+            cache = caches[key] = ThemeCache()
+        return cache
+
+    def getCacheKey(self, theme=None):
+        if not theme:
+            theme = self.getCurrentTheme()
+        key = "%s::%s" % (getSite().absolute_url(), theme)
+        return key
+
+    def getCacheStorage(self):
+        return self.__class__.CACHE_STORAGE
+
+    def invalidateCache(self):
+        """When our settings are changed, invalidate the cache on all zeo clients
+        """
+        self.__class__.CACHE_STORAGE = {}
+
+    def get_theme(self):
+        """Managing the theme cache is a plone.app.theming policy
+        decision. Moved out out Products.CMFPlone."""
+        cache = self.getCache()
+        themeObj = cache.themeObj
+        if not themeObj:
+            theme = self.getCurrentTheme()
+            themeObj = utils.getTheme(theme)
+            self.set_theme(theme, themeObj)
```

Comment:
I don't see an easy way around that. My use case is, that the fallback theme (barceloneta) is generated on the first hit on cms.localhost:8080. This is what causes the write-on-read errors and why I moved away from v_volatile.